stan/math/rev/fun/to_ad_stack.hpp

```diff
@@ -0,0 +1,84 @@
+#ifndef STAN_MATH_REF_FUN_TO_AD_STACK_HPP
+#define STAN_MATH_REF_FUN_TO_AD_STACK_HPP
+
+#include <stan/math/rev/meta.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * Converts given argument into a type that has any dynamic allocation on AD
+ * stack.
+ *
+ * For scalars this is a no-op.
+ * @tparam type of scalar
+ * @param a argument
+ * @return argument
+ */
+template <typename T, require_not_container_t<T>* = nullptr>
+T to_AD_stack(T&& a) {
+  return std::forward<T>(a);
+}
```

Comment:
You need `decltype(auto)` for forwarding here right?

```suggestion
template <typename T, require_not_container_t<T>* = nullptr>
decltype(auto) to_AD_stack(T&& a) {
  return std::forward<T>(a);
}
```

Reply:
Nope. That is needed only if you need to compiler to deduce return type. I know my return type is T (which can be a value or a lvalue reference).

I modified this not to use perfect forwarding anyway.

Reply:
Cool cool, for clarification I thought if you wanted to return an rvalue from a forwarded input you had to use `decltype(auto)` like the below? It looks like `T` makes a copy (though this function will almost surely be inlined anyway)

https://godbolt.org/z/hnTGe8

Reply:
Yeah it is a copy until RVO comes into play (and I think `decltype(auto)` works the same). There is also an option of `T&&`, but that has some other issues.

Reply:
icic I figured it might be copy ellision